systemd_info - add wildcards support

NomakCooper opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR adds support for wildcard expressions.

Examples:

```yaml
# Gather info using wildcards/expression
- name: Gather info of units that start with 'systemd-'
  community.general.systemd_info:
    unitname:
      - 'systemd-*'
  register: results
```

- Added new line in main description.
- Added new line in `unitname` description
- Added new example
- Added new integration tasks to test wildcard expressions and duplicates
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->
systemd_info
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

NomakCooper commented:
I have added commits with some changes.

- Using wildcard expressions in a task increases its execution time, so I had to make changes to the code to make it as fast as possible.

- I made changes for better management to make future maintenance of the code easier.

- The changes made do not change the usage or the result of the module.

russoz reviewed (commented):
hi @NomakCooper, thanks for the PR! :-) Couple of initial comments.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -235,30 +245,88 @@ def determine_category(unit):
+    if not category:
+        module.fail_json(msg="Could not determine the category for unit '{0}'.".format(unit))
+
+    base_props = {
+        'service': ['FragmentPath', 'UnitFileState', 'UnitFilePreset', 'MainPID', 'ExecMainPID'],
+        'target': ['FragmentPath', 'UnitFileState', 'UnitFilePreset'],
+        'socket': ['FragmentPath', 'UnitFileState', 'UnitFilePreset'],
+        'mount': ['Where', 'What', 'Options', 'Type']
+    }
```
Oops. I thought I had fixed all occurrences to avoid repetitions, but I missed something as always.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -267,92 +335,64 @@ def main():
-            ls = unit_data.get("loadstate", "").lower()
-            if ls not in ("not-found", "masked"):
-                fact.update(extract_unit_properties(unit_data, full_props))
+        all_units = list(units_info.keys())
+        resolved_units, non_matching = process_wildcards(selected_units, all_units, module)
+        units_to_process = sorted(resolved_units.keys())
+    else:
+        units_to_process = list(units_info.keys())
```
I made the same mistake as in the first PR. **`units_info`** and **`resolved_units`** already give the keys by default, so **`.keys()`** isn't needed.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -20,13 +20,15 @@
       - List of unit names to process.
       - It supports C(.service), C(.target), C(.socket), and C(.mount) units type.
-      - Each name must correspond to the full name of the C(systemd) unit.
+      - Each name must correspond to the full name of the C(systemd) unit or to a wildcard expression like V('ssh*') and V('*.service').
```
It is definitely useful for those reading the documentation to know that wildcard expressions are only supported from version **`10.5.0`**.

However, while searching the documentation, I didn’t find a specific semantic markup to indicate a version, so I think a simple **`C(10.5.0)`** will have to be used.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -235,30 +245,88 @@ def determine_category(unit):
+    non_matching_patterns = []
+
+    for pattern in selected_units:
+        matches = [unit for unit in all_units if fnmatch.fnmatch(unit, pattern)]
```
For the use of **`fnmatch`**, I searched in other modules of the collection, and at the moment, none of them use **`fnmatch.filter`** ( even though it has been available since **`Python 2.2`** ).

While searching, I found that, as of now, only the **`ansible.builtin.apt`** module uses **`fnmatch.filter`**.

Using **`fnmatch.fnmatch`** when **`fnmatch.filter`** is available doesn't make much sense, so I’ll try using it.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -267,92 +335,64 @@ def main():
+        "--no-legend"
+    ]
+
+    list_output = run_command(module, list_cmd)
```
how did I overlook the possibility that the host might have a different locale?

The output of **`systemctl`** commands depends on the locale, so if used in environments with specific languages, the parsing functions might fail.

NomakCooper commented:
@russoz thank you so much for the comments and suggestions.

I've made some changes:

- Removed redundancies in **`base_props`** and introduced the new **`get_category_base_props()`**
- Removed **`.keys()`** from the **`units_info`** and **`resolved_units`** lists.
- Replaced the more common **`fnmatch.fnmatch`** with **`fnmatch.filter`**.
- Replaced **`run_command`** with **`CmdRunner`** and set locale to **`en_US`**.
- Added a new line in the **`unitname`** description specifying that wildcard expressions are supported from **`community.general 10.5.0`**.

By introducing these changes, I aimed to make the code more readable and easier to maintain.

Replacing **`run_command`** with **`CmdRunner`** gave me some issues because the module uses three different **`systemctl`** syntaxes. 
So, I decided to create three separate functions that use **`CmdRunner`**. This way, however, the **`main()`** function is cleaner.

NomakCooper commented:
@russoz 

In the open issues, I found #4896, which was opened as a **Feature Idea** in 2022.

What is described in the issue is currently handled by **`systemd_info`**, except for **`timers`**.

I can use this PR to also extend support for the timer category with the **`base_props`** **`FragmentPath`**, **`UnitFileState`**, and **`UnitFilePreset`**

russoz reviewed (commented):
Hi @NomakCooper 

`CmdRunner` is meant to make us not repeat ourselves all the time. Please see how it is used (`pipx`, `xfconf` and `gconftool2` are good examples).

I will have some improvement suggestions on the argformats as well, but that's easier with one consolidated runner.

NomakCooper commented on plugins/modules/systemd_info.py:
```diff
@@ -195,11 +205,70 @@
+
+
+def run_version(module, systemctl_bin):
+    runner = CmdRunner(
+        module,
+        command=systemctl_bin,
+        arg_formats=dict(
+            version=cmd_runner_fmt.as_fixed("--version")
+        ),
+        default_args_order="version",
+        environ_update={'LC_ALL': 'en_US.UTF-8', 'LANG': 'en_US.UTF-8'},
+        force_lang="en_US",
+        check_rc=True,
+    )
+    with runner("version") as ctx:
+        rc, stdout, stderr = ctx.run()
+    return stdout.strip()
+
+
+def list_units(module, systemctl_bin):
+    runner = CmdRunner(
+        module,
+        command=systemctl_bin,
+        arg_formats=dict(
+            list_units=cmd_runner_fmt.as_fixed("list-units"),
+            no_pager=cmd_runner_fmt.as_fixed("--no-pager"),
+            typ=cmd_runner_fmt.as_fixed("--type"),
+            types=cmd_runner_fmt.as_fixed("service,target,socket,mount"),
+            all=cmd_runner_fmt.as_fixed("--all"),
+            plain=cmd_runner_fmt.as_fixed("--plain"),
+            no_legend=cmd_runner_fmt.as_fixed("--no-legend"),
+        ),
```
Instead of creating a dedicated **`module_utils`** with a runner, I could create a **`base_runner`** inside the module with all possible arguments. Then, the functions **`run_version()`**, **`list_units()`**, and **`show_unit_properties()`** would use **`base_runner`** as ctx, passing the args to use, the units, and the properties.

like:

```python
# Create one base runner
base_runner = CmdRunner(
    module,
    command=systemctl_bin,
    arg_formats={
        'version':         cmd_runner_fmt.as_fixed("--version"),
        'list_units':      cmd_runner_fmt.as_fixed("list-units"),
        'no_pager':        cmd_runner_fmt.as_fixed("--no-pager"),
        'typ':             cmd_runner_fmt.as_fixed("--type"),
        'types':           cmd_runner_fmt.as_fixed("service,target,socket,mount"),
        'all':             cmd_runner_fmt.as_fixed("--all"),
        'plain':           cmd_runner_fmt.as_fixed("--plain"),
        'no_legend':       cmd_runner_fmt.as_fixed("--no-legend"),
        'p':               cmd_runner_fmt.as_fixed("-p"),
        'props':           cmd_runner_fmt.as_list(),
        'dashdash':        cmd_runner_fmt.as_fixed("--"),
        'unit':            cmd_runner_fmt.as_list(),
    },
    default_args_order="version list_units no_pager typ types all plain no_legend p props dashdash unit",
    check_rc=True,
)

# list_units with base_runner
def list_units():
    context = "list_units no_pager typ types all plain no_legend"
    with base_runner(context) as ctx:
        rc, stdout, stderr = ctx.run()
    return stdout.strip()
```

NomakCooper replied:
after looking at the **`pipx`** module, I wrote **`plugins/module_utils/systemd_info.py`** with a dedicated runner that includes the possible **`systemctl`** arguments used by **`systemd_info`**. I then updated the module to use **`systemd_info_runner`**.

like:
```python
def run_version(runner):
    with runner("version") as ctx:
        rc, stdout, stderr = ctx.run()
    return stdout.strip()


def list_units(runner):
    context = "list_units no_pager typ types all plain no_legend"
    with runner(context) as ctx:
        rc, stdout, stderr = ctx.run()
    return stdout.strip()


def show_unit_properties(runner, prop_list, unit):
    props_str = ",".join(prop_list)
    context = "show p props dashdash unit"
    with runner(context) as ctx:
        rc, stdout, stderr = ctx.run(props=props_str, unit=unit)
    return stdout.strip()
```

NomakCooper commented:
@russoz as you suggested, I changed the **`CmdRunner`**.

I introduced a new **`plugins/module_utils/systemd_info.py`** with a dedicated runner that includes all the args used by **`systemd_info`**.

In the module, functions now call the runner, passing arguments like units, props, types, etc.

For the unit **`types`**, I used **`cmd_runner_fmt.as_func`**. 
This way, the supported categories can be managed directly in the module without needing to modify the runner.

I also removed **`environ_update`** and **`force_lang`**

russoz commented on plugins/module_utils/systemd_info.py:
If I  may suggest, name this file `systemd.py`, so that any other module related to systemd might benefit from that in the future.

NomakCooper replied:
I was thinking of implementing it just for this module, but if it can be useful to other contributors, then sure.

NomakCooper replied:
Should I also rename **`systemd_info_runner`** ( to **`systemd_runner`** ) in **`plugins/module_utils/systemd.py`** ?

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -195,11 +205,29 @@
+from ansible_collections.community.general.plugins.module_utils.systemd_info import systemd_info_runner
+
+
+def run_version(runner):
```
minor nit: wouldn't `get_version()` depict a more accurate description of what the function does? I mean, it is plain to see, but as far as naming goes... :-)

NomakCooper replied:
Yes, maybe that's better, more intuitive. I got a bit too carried away by cmd(run)ner. 😄

NomakCooper commented:
@russoz done

- The runner is no longer specifically tied to **`systemd_info`**; I’ve renamed both the file and the runner.
- Changed **`run_version()`** to **`get_version()`**.

For the issue https://github.com/ansible-collections/community.general/issues/4896 I previously linked in the comments, what do we want to do?

I'm not sure if it's possible to use this PR to also add support for **`timer`** units.
Also, I’m not sure if a review from Felix is needed 🙂

Obviously, thanks as always!

russoz commented:
Hi, is there anything different about thw timers, in terms of the general logic of the module? Does it require creating additional logic, or is it just creating the new category and parameters?

NomakCooper commented:
Timer units have various properties, but many of them are variable since they depend on how the timers are configured.

However, every timer unit shares the basic properties already present in the module for **`target`** and **`socket`**.

So, it would be enough to add a new entry in **`base_props`**, adjust **`unit_types`** list, **`determine_category()`**, and, of course, update the **`documentation`** and **`changelogs/fragments`**.

If desired, one or two additional tasks could be added to the **`integration`** test to verify **`timer`** units.

NomakCooper commented:
Additionally, there's another thing we forgot. 
Since we added the new **`module_utils/systemd.py`**, I think it's necessary to update **`.github/BOTMETA.yml`** as well.

russoz commented:
> Timer units have various properties, but many of them are variable since they depend on how the timers are configured.
> 
> However, every timer unit shares the basic properties already present in the module for **`target`** and **`socket`**.
> 
> So, it would be enough to add a new entry in **`base_props`**, adjust **`unit_types`** list, **`determine_category()`**, and, of course, update the **`documentation`** and **`changelogs/fragments`**.
> 
> If desired, one or two additional tasks could be added to the **`integration`** test to verify **`timer`** units.

I am happy with the PR as-is (well, I am about to make another suggestion on the arg_formats, but minor stuff), so it is kind of your call to make: would you prefer to get this one merged first and then raise a second PR, or would you rather put that extra feature in this PR as well?

NomakCooper commented:
In this PR, many changes have already been made, so I prefer to keep this PR focused on wildcard expression support and extend timer unit support in a new PR.

russoz commented on plugins/module_utils/systemd.py:
```diff
@@ -0,0 +1,37 @@
+def systemd_runner(module, command, **kwargs):
+    arg_formats = dict(
+        version=cmd_runner_fmt.as_fixed("--version"),
+        list_units=cmd_runner_fmt.as_fixed("list-units"),
+        no_pager=cmd_runner_fmt.as_fixed("--no-pager"),
```
These (and others, whenever it makes sense) can be combined as:
```suggestion
        list_units=cmd_runner_fmt.as_fixed(["list-units", "--no-pager"]),
```

russoz commented on plugins/module_utils/systemd.py:
```diff
@@ -0,0 +1,37 @@
+        plain=cmd_runner_fmt.as_fixed("--plain"),
+        no_legend=cmd_runner_fmt.as_fixed("--no-legend"),
+        show=cmd_runner_fmt.as_fixed("show"),
+        p=cmd_runner_fmt.as_fixed("-p"),
+        props=cmd_runner_fmt.as_list(),
```
The `p` arg is kinda artificial, it is actually about the `props`. In that case, it might be better off having one format:
```suggestion
        props=cmd_runner_fmt.as_func(lambda v: [] if not v else ["-p", ",".join(v)]),
```
and then ...

russoz commented on plugins/modules/systemd_info.py:
```diff
@@ -195,11 +205,29 @@
+
+
+def show_unit_properties(runner, prop_list, unit):
+    props_str = ",".join(prop_list)
+    context = "show p props dashdash unit"
+    with runner(context) as ctx:
+        rc, stdout, stderr = ctx.run(props=props_str, unit=unit)
```
... here it would become:
```suggestion
    context = "show props dashdash unit"
    with runner(context) as ctx:
        rc, stdout, stderr = ctx.run(props=props_list, unit=unit)
```

russoz reviewed (commented):
Hi there, as mentioned, a couple of suggestions on the arg formats, but even without these adjustments, LGTM

NomakCooper commented:
> Hi there, as mentioned, a couple of suggestions on the arg formats, but even without these adjustments, LGTM

OK It makes sense., this way some args are combined since they are related to each other.
I'll try making these changes and running the tests.

While I'm at it, I'll also update the **`changelogs/fragments`** with a clearer and more precise note.

Do I also need to add **`module_utils/systemd`** to **`.github/BOTMETA.yml`**?

russoz commented:
> OK It makes sense., this way some args are combined since they are related to each other. I'll try making these changes and running the tests.
> 
> While I'm at it, I'll also update the **`changelogs/fragments`** with a clearer and more precise note.

Awesome, thanks! :-)

> Do I also need to add **`module_utils/systemd`** to **`.github/BOTMETA.yml`**?

Yes. That file basically drives the bot, so whenever some issu or PR mention that file, the bot will mention your GH user, so that you get notified.

NomakCooper commented:
@russoz updated, the module works well, and the integration test ran without any issues.

- Combined the **`arg_formats`** and updated **`list_units()`** and **`show_unit_properties()`**.
- Improved the note in **`changelogs/fragments`**.
- Added a new entry in **`BOTMETA.yml`** for **`module_utils/systemd.py`**.

NomakCooper commented on plugins/module_utils/systemd.py:
```diff
@@ -0,0 +1,35 @@
+    arg_formats = dict(
+        version=cmd_runner_fmt.as_fixed("--version"),
+        list_units=cmd_runner_fmt.as_fixed(["list-units", "--no-pager"]),
+        typ=cmd_runner_fmt.as_fixed("--type"),
+        types=cmd_runner_fmt.as_func(lambda v: [] if not v else [",".join(v)]),
```
They could become a single arg.

like:
```python
types=cmd_runner_fmt.as_func(lambda v: [] if not v else ["--type", ",".join(v)]),
```

russoz reviewed (approved):
LGTM

felixfontein commented:
@NomakCooper thanks for your contribution!
@russoz thanks for reviewing!
